### ISSUE TITLE
push: Unsupported pickle protocol: 5 on python 3.7.11

### ISSUE THREAD

**Issue by jnissin** [iterative/dvc#7726]
# Bug Report

## Description

When attempting to push to remote (Google Cloud Storage) using DVC version `2.10.2` and Python version `3.7.11` I encounter the following error:

```console
$ dvc push     
ERROR: unexpected error - unsupported pickle protocol: 5              

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```

In my tests on python 3.8 this issue does not happen.

### Reproduce

Install DVC version `2.10.2` with Python version `3.7.11` and attempt to push to remote.

### Expected

Expected to pickle successfully and push the pickled data to remote (Google Cloud Storage).

### Environment information

```console
$ dvc doctor
DVC version: 2.10.2 (conda)
---------------------------------
Platform: Python 3.7.11 on Darwin-21.2.0-x86_64-i386-64bit
Supports:
        gs (gcsfs = 2022.1.0),
        webhdfs (fsspec = 2022.1.0),
        http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
        https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6)
Cache types: reflink, hardlink, symlink
Cache directory: apfs on /dev/disk1s1s1
Caches: local
Remotes: gs
Workspace directory: apfs on /dev/disk1s1s1
Repo: dvc, git
```

**Comment by skshetry** [comment]
@jnissin, can you share the output from `dvc push -v`?

**Comment by jnissin** [comment]
Here you go:

```console
$ dvc push -v
2022-05-11 13:48:51,201 DEBUG: Preparing to transfer data from '/Users/joonas.nissinen@futurice.com/Documents/code/work/yle/valimo/.dvc/cache' to 'gs://valimo-dvc-storage'
2022-05-11 13:48:51,201 DEBUG: Preparing to collect status from 'gs://valimo-dvc-storage'
2022-05-11 13:48:51,201 DEBUG: Collecting status from 'gs://valimo-dvc-storage'
2022-05-11 13:48:51,204 ERROR: unexpected error - unsupported pickle protocol: 5
------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/cli/__init__.py", line 90, in main
    ret = cmd.do_run()
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/cli/command.py", line 22, in do_run
    return self.run()
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/commands/data_sync.py", line 68, in run
    glob=self.args.glob,
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/repo/__init__.py", line 48, in wrapper
    return f(repo, *args, **kwargs)
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/repo/push.py", line 69, in push
    obj_ids, jobs, remote=remote, odb=odb or dest_odb
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/data_cloud.py", line 91, in push
    cache_odb=self.repo.odb.local,
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/data/transfer.py", line 155, in transfer
    src, dest, obj_ids, check_deleted=False, jobs=jobs, **kwargs
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/data/status.py", line 181, in compare_status
    dest, obj_ids, index=dest_index, jobs=jobs, **kwargs
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/data/status.py", line 138, in status
    odb, index, dir_objs, name, cache_odb, jobs=jobs
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/data/status.py", line 36, in _indexed_dir_hashes
    indexed_dirs = set(index.dir_hashes())
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/data/db/index.py", line 119, in dir_hashes
    yield from (hash_ for hash_, is_dir in self.index.items() if is_dir)
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/dvc/data/db/index.py", line 119, in <genexpr>
    yield from (hash_ for hash_, is_dir in self.index.items() if is_dir)
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/_collections_abc.py", line 744, in __iter__
    yield (key, self._mapping[key])
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/diskcache/persistent.py", line 730, in __getitem__
    return self._cache[key]
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/diskcache/core.py", line 1232, in __getitem__
    value = self.get(key, default=ENOVAL, retry=True)
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/diskcache/core.py", line 1173, in get
    value = self._disk.fetch(mode, filename, db_value, read)
  File "/opt/anaconda3/envs/valimo-local/lib/python3.7/site-packages/diskcache/core.py", line 285, in fetch
    return pickle.load(io.BytesIO(value))
ValueError: unsupported pickle protocol: 5
------------------------------------------------------------
2022-05-11 13:48:51,316 DEBUG: Removing '/Users/joonas.nissinen@futurice.com/Documents/code/work/yle/.mJCQTQ7VWcdKbQvxRdcp2y.tmp'
2022-05-11 13:48:51,318 DEBUG: Removing '/Users/joonas.nissinen@futurice.com/Documents/code/work/yle/.mJCQTQ7VWcdKbQvxRdcp2y.tmp'
2022-05-11 13:48:51,318 DEBUG: Removing '/Users/joonas.nissinen@futurice.com/Documents/code/work/yle/.mJCQTQ7VWcdKbQvxRdcp2y.tmp'
2022-05-11 13:48:51,318 DEBUG: Removing '/Users/joonas.nissinen@futurice.com/Documents/code/work/yle/valimo/.dvc/cache/.68G3bcZEzc4zjdHpDsNqYj.tmp'
2022-05-11 13:48:51,322 DEBUG: Version info for developers:
DVC version: 2.10.2 (conda)
---------------------------------
Platform: Python 3.7.11 on Darwin-21.2.0-x86_64-i386-64bit
Supports:
        gs (gcsfs = 2022.1.0),
        webhdfs (fsspec = 2022.1.0),
        http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
        https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6)
Cache types: reflink, hardlink, symlink
Cache directory: apfs on /dev/disk1s1s1
Caches: local
Remotes: gs
Workspace directory: apfs on /dev/disk1s1s1
Repo: dvc, git

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
2022-05-11 13:48:51,326 DEBUG: Analytics is enabled.
2022-05-11 13:48:51,457 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/var/folders/4p/0n_mm8js23x1vhydv_3yq8b40000gq/T/tmp_4ld4dv3']'
2022-05-11 13:48:51,459 DEBUG: Spawned '['daemon', '-q', 'analytics', '/var/folders/4p/0n_mm8js23x1vhydv_3yq8b40000gq/T/tmp_4ld4dv3']'
```

**Comment by daavoo** [comment]
This seems like https://dvc.org/doc/user-guide/troubleshooting#pickle but for some reason we are not catching the exception in `pull` in order to redirect users to the link:

https://github.com/iterative/dvc/blob/47d10c4a8b9bdc18dec17a9380f0593e43383c4c/dvc/utils/decorators.py#L13-L30

**Comment by jnissin** [comment]
True, I can confirm that the following fixed the issue:

```
rm -r .dvc/tmp/index .dvc/tmp/md5s .dvc/tmp/links
```

This is probably due to the fact that we have python 3.8 running in our pre-commit hooks environment.

**Comment by pmrowla** [comment]
@jnissin this behavior (but not the incorrect error message) is expected if you are mixing python3.8 and 3.7 usage within the same repo. See the docs linked by @daavoo, you will need to remove `.dvc/tmp/index` before running in python 3.7 in this case.

**Comment by jnissin** [comment]
Alright, I'll try to switch our pre-commit environment to use python 3.7 as well. Thanks a lot!

**Comment by efiop** [comment]
@jnissin Btw, did you migrate from an older dvc version recently? We had a bug where we didn't set pickle version https://github.com/iterative/dvc/pull/7222 , but it was fixed 5 months back in 2.9.4
